Doc/#1209 - #1250 - Add documentation page for each application and page template (#1263)
Avaiga/taipy-doc: docs/tp_templates/applications/sdm_app.md

> Building a Taipy application with multiple complex scenarios usually requires a
> significant amount of time and effort, in particular at the beginning of a project
> when the user interface requirements are not completely stable. This can delay the
> time-to-market and increase development costs.
> 
> To reduce the development time, Taipy provides a scenario management application template,
> which leverages Taipy visual elements to speed up bootstrapping a standard application.
> 
> Out-of-the-box, the scenario management application is a multi-page Taipy application
> that allows the user to visualize and manage scenarios and data nodes. The user can select a
> scenario, view its directed acyclic graph (DAG), manage its data nodes, upload data files,
> and submit the scenario for execution.

Reviewer: ```suggestion
The scenario management template provides a foundational structure for applications
requiring scenario-based analysis and data management. It's designed to help users create, 
manage, and compare different scenarios to facilitate decision-making and what-if
analyses. It is ideal for applications in forecasting, simulation, and optimization, where
multiple scenarios need to be evaluated and compared.

Out of the box, the template provides a multi-page application structure with two pages predefined:
- A scenario page to select, visualize, submit, analyze, and manage scenarios and data nodes. 
- A job page to monitor and manage submissions and jobs.
```

<figure>
  <img src="../img/sdm_app_template_with_scenario_dark.jpeg" class="visible-dark" />
  <img src="../img/sdm_app_template_with_scenario_light.jpeg" class="visible-light"/>
  <figcaption>Taipy scenario management application created by the scenario management template out-of-the-box.</figcaption>
</figure>

The application also includes a job monitoring page to track the status of the submitted jobs.

<figure>
  <img src="../img/sdm_app_template_with_job_dark.jpeg" class="visible-dark" />
  <img src="../img/sdm_app_template_with_job_light.jpeg" class="visible-light"/>
  <figcaption>Manage job with the Taipy scenario management application created by the scenario management template out-of-the-box.</figcaption>

> 

Reviewer: ```suggestion
</figure>
```

> This template offers several key benefits:
> 
> - **Accelerated Development**: By leveraging scenario and data management visual elements,
>     developers can quickly bootstrap a standard application, saving significant
>     development time to focus on delivering business value more efficiently.
> - **Comprehensive Features**: The template provides a best-practice folder scaffolding for a
>     scenario management application, which supports a wide range of functionalities, including
>     data and scenario visualization, job orchestration monitoring, providing a comprehensive
>     solution for various use cases.
> - **Customization**: The generated application is highly customizable with pre-built
>     placeholders for various components, allowing developers to tailor the application
>     to meet specific requirements and use cases.
> - **Deployment-Ready**: The template provides Docker deployment helpers for both development purpose
>     and production environment.

Reviewer: ```suggestion

The template provides support for:
- Multi-pages structure
- Scenario and data management
- Job monitoring
- gitignore setup
- Docker file setup
```

```console
$ taipy create --application sdm
[1/6] Application root folder [taipy_application]:
[2/6] Application main Python file [main.py]:
[3/6] Application title [Taipy Application]:
[4/6] With TOML Config? (No):
[5/6] With a new Git repository? (No):
[6/6] Select With Docker deployment
    1 - No
    2 - For development
    3 - For production
    Choose from [1/2/3] (1):
New Taipy application has been created at ./taipy_application

To start the application, change directory to the newly created folder:
    cd ./taipy_application

You can then run the application as follows:
    taipy run main.py
```

??? info "Default answers"

    In the CLI, the default answer for each question is displayed in the square brackets.
    You can provide an answer or press Enter to use the default value.

Each question in the CLI corresponds to a specific aspect of the application. The following
sections describe each question in detail.

!!! note "Available in Taipy Enterprise edition"

    Questions 6 is only relevant to the [Taipy Enterprise Edition](https://taipy.io/enterprise)

    [Contact us](https://taipy.io/book-a-call){: .tp-btn .tp-btn--accent target='blank' }

> ## 1. Application root folder

Reviewer: I would turn the title into an enumeration to make the TOC lighter.

Reviewer 2: No authorization here?

Author: I created a new issue for this: https://github.com/Avaiga/taipy-enterprise/issues/642

- Specifies the root folder of the application.
- The default value is "taipy_application".

## 2. Application main Python file

- Sets the name of the main Python file (entry point) of the application.
- The default value is "main.py".

## 3. Application title

- Specifies the title displayed in the web application.
- The default value is "Taipy Application".

## 4. With TOML Config

- Specifies whether the application will use TOML configuration files, otherwise the configuration

>   will be in a Python file.

Reviewer 2: ```suggestion
  will be in generated in a Python script file.
```

- The default value is "No".

## 5. With a new Git repository

> - Specifies whether the application will be initialized as a new Git repository.

Reviewer 2: ```suggestion
- Specifies whether the application directory should be initialized as a new Git repository.
```

- The default value is "No".

## 6. With Docker deployment

- Specifies Docker support for the application.
- Options:
    - "No": No Docker support.
    - "For development": Add a minimal version of `Dockerfile` and `docker-compose.yml` for development.
    - "For production": Add a production-ready `Dockerfile` and `docker-compose.yml`.
- The default value is "No".

# Application description

The generated application has the following folder structure:

```plaintext
taipy_application/
├──── algos/
│   ├──── __init__.py
│   └──── algos.py
│
├──── config/
│   ├──── __init__.py
│   └──── config.py
│
├──── pages/
│   ├──── __init__.py
│   ├──── job_page/
│   ├──── scenario_page/
│   └──── root.py
│

> ├──── .taipyignore

Reviewer: missing the gitignore, no?

Reviewer 2: No auth?

├──── docker-compose.yml
├──── Dockerfile
├──── main.py
└──── requirements.txt
```

Your application's folder structure may vary depending on the options you selected during the
creation process. Here is a brief overview of the key components:

- *algos/*: Contains the *algos.py* file, designed to contain various Python functions used to
    configure tasks.
- *config/*: Contains the *config.py* file, where you can put your application's configuration.

>     The configuration will be imported into the main application file.

Reviewer 2: ```suggestion
    The configuration will be imported by the main application file.
```

- *pages/*: Contains the application pages.
    - *root.py* is the root page of the application, which layouts the application.
        It includes a navigation bar, and a sidebar with the
        [scenario_selector](../../refmans/gui/viselements/corelements/scenario_selector.md) and
        [data_node_selector](../../refmans/gui/viselements/corelements/data_node_selector.md)
        visual elements.
    - *job_page/* contains the page for job monitoring using the
        [job_selector](../../refmans/gui/viselements/corelements/job_selector.md) visual element.
    - *scenario_page/* contains the page for scenario analysis and data management. It
        shows a scenario, its DAG, and its data nodes. It uses the
        [scenario](../../refmans/gui/viselements/corelements/scenario.md), the
        [scenario_dag](../../refmans/gui/viselements/corelements/scenario_dag.md)
        and the [data_node](../../refmans/gui/viselements/corelements/data_node.md)
        visual elements.

> - *.taipyignore*: Specifies files to be protected when running the web server. Please refer to the

Reviewer: We are missing the .gitignore here.

    [Protect private files](../../userman/run-deploy/run/protect_files.md) page for more
    information.
- *docker-compose.yml* and *Dockerfile*: The Docker configuration for building and running the
    application as a Docker container.
- *main.py*: The main Python file of the application.
- *requirements.txt*: Contains the Python dependencies required by the application.

> # Customizing the application

Reviewer: ```suggestion
# How to customize the application
```

Everything in the generated application can be updated to fit your needs. It includes
the Python code, the configuration files, and any other resources. Specifically, in
the **algos/** and **config/** folders, there are placeholders that you can customize
precisely to your use case.

> ## Customizing the tasks' functions

Reviewer: ```suggestion
## Custom tasks

> The *algos/* folder contains the *algos.py* file, designed to contain various Python

Reviewer 2: So the convention is to have file/dir names in typeset (tt) and variables/functions in italics (i).
All this section does the exact opposite.

functions used to configure tasks. The functions are strongly use-case dependent,
for example, cleaning data, performing analysis, or running simulations, among other
tasks.

> In the *algos.py* file, the `clean_data()` method is provided as a placeholder. You can

Reviewer 2: For example:
```suggestion
In the `algos.py` file, the *clean_data()* function is provided as a placeholder. You can
```
But... about something else
is *clean_data()* a method??

> edit/replace the placeholder method `clean_data()` and add your tasks' functions here.

Author: ```suggestion
edit or replace the function body and add your tasks' functions here.
```

```python title="algos/algos.py"
def clean_data(df, replacement_type):
    df = df.fillna(replacement_type)
    return df

def analyze_data(df):
    ...
    return analysis_results
```

To configure tasks using these functions, make sure to update the import statements in
the *algos/\_\_init\_\_.py* file.

```python title="algos/__init__.py"
from .algos import clean_data, analyze_data
```

Then when [customizing the configuration](#customizing-the-configuration), you can
import these functions and use them to configure various tasks for your application.

??? note "Renaming generated folders or files"

    If you decide to rename the *algos.py* file or the *algos/* folder, make sure to
    update the import statements in the *algos/\_\_init\_\_.py* file and any other
    import statements.

> ## Customizing the configuration

Reviewer: ```suggestion
## Custom configuration
```

> The *config/config.py* file contains the `configure()` method, which will be called in

Reviewer 2: ```suggestion
The *config/config.py* file contains the `configure()` function, which will be called in
```

the main application file to configure the application.

If your answer to the ["With TOML Config"](#4-with-toml-config) question was "No"

> (as default), the placeholder configuration will be stored in the `configure()` method

Reviewer 2: ```suggestion
(the default), the placeholder configuration will be stored in the `configure()` function
```

[as Python code](../../userman/advanced_features/configuration/advanced-config.md#python-code-configuration). 
Import the added functions from the *algos/algos.py* file and use them to configure the tasks.

```python title="config/config.py"

from algos import clean_data, analyze_data

def configure():
    ...
```

If your answer to the ["With TOML Config"](#4-with-toml-config) question was "Yes",

> the placeholder configuration will be stored in `config.toml` file. Update the

Reviewer 2: See? Here it is correct. 'config.toml' is a file name indeed.

`config.toml` file to configure the application to your specific use case, or you can
use [Taipy Studio](../../userman/ecosystem/studio/index.md) for generating a TOML

> file that can be loaded in the `configure()` method.

Reviewer 2: ```suggestion
file that can be loaded in the `configure()` function.
```

!!! note "Loading the TOML configuration"

    If you have a different name for the TOML file, make sure to update the file name in the
    `configure()` method.

>  ## Customizing the pages

Reviewer: ```suggestion
 ## Custom pages
```

content of the pages to fit your specific requirements, as well as customize
the grid layout of the application and the visual elements.

*pages/root.py* defines the layout of the application, including the navigation bar,
and a sidebar. Here, you can customize the [layout](../../refmans//gui/viselements/generic/layout.md),
the [scenario_selector](../../refmans/gui/viselements/corelements/scenario_selector.md), and
the [data_node_selector](../../refmans/gui/viselements/corelements/data_node_selector.md).

In *pages/job_page/job_page.py*, you can customize the
[job_selector](../../refmans/gui/viselements/corelements/job_selector.md) visual element, which
lists all jobs of the application and allows users to select and manage them.

In *pages/scenario_page/scenario_page.py*, you can customize the
[scenario](../../refmans/gui/viselements/corelements/scenario.md),
the [scenario_dag](../../refmans/gui/viselements/corelements/scenario_dag.md), and the
[data_node](../../refmans/gui/viselements/corelements/data_node.md) visual elements to display
the information of the selected scenario and data node, as well as modifying the

> `notify_on_submission()` method to handle the notification of a scenario's submission.

Reviewer 2: ```suggestion
`notify_on_submission()` function to handle the notification of a scenario's submission.
```

# How to run the application

To run the application, change to the newly created folder and run the application using
`taipy run main.py`.

> ```console

Reviewer 2: sh?

$ cd ./taipy_application
$ taipy run main.py
```

If the newly created application supports Docker, you can also run the application using `docker-compose`.

> ```console

Reviewer 2: sh?
